### ISSUE TITLE
[Report] The bot bans users, but it does not unban them after

### ISSUE THREAD

**Issue by LawrenceJGD** [J-Rios/TLG_JoinCaptchaBot#157]
Multiple user of several groups are being banned instead of being kicked when they don't solve the captcha since some time ago, and we don't know why.

![join_captcha_bot_bans](https://user-images.githubusercontent.com/11948239/167933344-cd4a31cb-0710-4f32-97a3-872d1513c5d4.png)
![join_captcha_bot_bans_2](https://user-images.githubusercontent.com/11948239/167933998-c4623c7c-0a43-4fd9-a0d0-2e4a1479c1a7.png)

**Comment by J-Rios** [comment]
Hello,

Let me add a couple of notes regarding kick/ban behavior to be in context:

1. Telegram doesn't have a "kick" method, just a "ban" (that remove the user from a group and also apply all restriction to them to disallow the user to join again). So the only way to handle kicks is by banning the user following by unbaning (remove all the restrictions that has been applied to the user at ban), and this is what Joincaptchabot do to kick.

2. Joincaptchabot just kicks an user when the captcha is not complete. However, if that user try again to join the same group and keeps not solving the captcha and being kicked, after multiple retries in a short time, the user is considered a Bot and then it is Ban.

3. Another Telegram rule is that any user removed from a group by an Admin (i.e. JoinCaptchaBot) can't be added to the group by another group member. If an user has been kicked, then it only can join the group by itself (no adding from another member or Admin is allowed).

Now, regarding your issue, I'm not sure that it's related to any of the previous points. However, the last time Bot kick/ban code was modified near two years ago and no one else has notify this issue from there (I also didn't see it) so it is really disconcerting.

Are you sure that all of this user are new users that didn't solve the captcha just 1 time (and not related to above point 2)?

Please take a look into Telegram group settings for "Recent Actions" when a user is kicked and give some feedback regarding that. Here it is an image of default and expected Bot behavior regarding kicking a user from this "recent actions":

![image](https://user-images.githubusercontent.com/12136967/167955199-4e9d469c-d8d8-45cb-9f94-92254acb75dd.png)


Regards.

**Comment by LawrenceJGD** [comment]
Sorry for taking so long to reply.

I confirmed that the bot bans even if the user enters for the first time in the chat, I made more screenshots to prove it:

![First screenshot](https://user-images.githubusercontent.com/11948239/170397987-69cd44bf-d12a-4ccb-8d48-8d595d1a595e.png)
![Second screenshot](https://user-images.githubusercontent.com/11948239/170397994-1e134cf8-6855-42a5-9ef8-c2cc432e97fe.png)
![Third screenshot](https://user-images.githubusercontent.com/11948239/170398005-4f305011-0c20-4fd6-8960-711eb8eb7924.png)
![Fourth screenshot](https://user-images.githubusercontent.com/11948239/170398020-16bd166c-74de-4c8b-a4eb-fd92fc04c171.png)
![Fifth screenshot](https://user-images.githubusercontent.com/11948239/170398025-0253a2de-11b9-4b98-bb2f-318482c5dfd1.png)
![Sixth screenshot](https://user-images.githubusercontent.com/11948239/170398038-96061a8a-529a-4a6e-85e9-6429f63ea79f.png)
![Seventh screenshot](https://user-images.githubusercontent.com/11948239/170398047-e48aa130-42bb-4bbf-85d6-75f6c62e62ec.png)
![Eighth screenshot](https://user-images.githubusercontent.com/11948239/170398058-bc957ce2-bc84-4f37-bf8c-b1f49a394b51.png)
![Ninth screenshot](https://user-images.githubusercontent.com/11948239/170398079-e3fef16f-9151-4ed6-864d-10d5265907f2.png)

Note: I censored the names but colored them, so you can see who does what.

In our chat, we use another Bot that's called GroupHelpBot, we activated an option that mutes people until they have an alias, maybe that's helping to cause the problem, but we had that option activated before the bug happened, so it would be strange.

**Comment by J-Rios** [comment]
Hi,

No problem, from the images above:

1. Once a user enters the group, the GroupHelpBot restrict all their rights (including sending messages, so they won't be able to solve the captcha). As you have said, this is intended for users without alias, so a note here that the fate of users without alias, are to be kicked out from the group because they can't write messages to solve the captcha.

2. Indeed, the remove restrictions when a user is kicked is missing from the recent actions history, so thinks that you are completely right and there is an issue.

3. JoinCaptchaBot is trying to kick the user when the issue occurs (the Bot is not trying to Ban).

4. JoinCaptchaBot is configured on the group to restrict media messages for 1 day to users that solve the captcha (/restrict_non_text enable is on).

So, seems a bug or unexpected behavior at some point (or with some specific configuration) when kicking an user.

Related lines of code where the issue should be are:
[On kick usage](https://github.com/J-Rios/TLG_JoinCaptchaBot/blob/2350036f57a7b027c872df175fa7035b37095322/sources/join_captcha_bot.py#L2993-L3027)
[On kick method](https://github.com/J-Rios/TLG_JoinCaptchaBot/blob/2350036f57a7b027c872df175fa7035b37095322/sources/tlgbotutils.py#L235-L253)

Of course, it could also be inside current used version of python-telegram-bot framework, so we need to check for bugs and changelogs to newer version.

I have something now for trying to replicate the issue, but would be nice if you can provide the next info (will be really helpful):

- The chat ID of the group where this happen, to allow me check Bot log on that group and also all the configuration of the Bot on it (you can get it using the command /chatid on the group).

- The issue happen with all members or just sometimes?

Best Regards

**Comment by LawrenceJGD** [comment]
Hi, I would like to avoid sending the chat ID here for privacy issues, but what I can do is tell you via Telegram the ID, I assume your alias is the one that appears when using the `/about` command.

And yes, the problem happens to everyone who tries to join the group, we have to unban people after the bot bans them to get them to try to join again.

By the way, is it ok if we speak in Spanish? That way it will be easier to talk.

**Comment by J-Rios** [comment]
Sure, you can contact me through Telegram on the /about alias, and we will continue the conversation there, in spanish :)

P.D. chat IDs are useless and meaningless for anyone that are not already in that group, you can't get any information of the group from it, so it can be considered safe-privacy enough. Any user on the group received the chat ID on each message, so even Telegram clients could shows the chat ID (of course official telegram clients doesnt shows it because it doesn't provides any functionality for common users).

**Comment by J-Rios** [comment]
Can't replicate the issue, I have revised the code and the logic seems ok, so I'm afraid that it could be something weird related to networking issue (i.e. Telegram Server receiving/responding to unrestrict request before the ban request is managed) or by some specific setup of your group that I can't replicate (Telegram Server makes distinction from private groups, public groups, supergroups, channels and also behaves differently depending on the number of members on a group, so it is difficult to test all use cases)...

Anyway, I have done some changes to check if any of this solves the issue:

1. I have updated the python-telegram-bot framework to latest stable version (just in case there was something related to it).

2. I have modified the telegram kick function to just use one request (seems that we can use the [unbanChatMember request with only_if_banned set to False](https://core.telegram.org/bots/api#unbanchatmember) to get a "kick" behavior, instead the 2 requests of ban+unban procedure that I was using. Hopefully this could fix any kind of issue here.

I will update the deployed Bots (@join_captcha_bot and @pro_join_captcha_bot) tomorrow with this fix. I will tell you once done, to request you to check if it has been solved.

See you.
